### ISSUE TITLE
Unable to switch between Humanoid 3D Models

### ISSUE THREAD

**Issue by jonathanchh1** [Rajawali/Rajawali#2311]
Hi, this occurs whenever i try to switch between two models, essentially, i am doing it by calling initScene method again and passing new instance of a new raw resource and whenever the obj parser begins to parse the object it tombstones. 
```
  private fun onStartRenderer() {
        try {
            renderer = Object3DRenderer(this)
            renderer?.sceneCachingEnabled = true
            renderer?.currentObject3D?.set(R.raw.obj_muscles_unfrozen)
            renderer?.let {
                surfaceView?.setSurfaceRenderer(it)
            }
            isSurfaceViewRendered.set(true)
            scaleGestureDetector = ScaleGestureDetector(this, ScaleListener())
            gestureDetector = GestureDetector(this, RotationListener())
        } catch (e: Exception) {
            Log.e("TAG", e.toString())
        }
    }
```

###   fun initialize(newScene : Scene){
        try {
            mLight = PointLight()
            mLight!!.setPosition(0.0, 0.0, 4.0)
            mLight!!.power = 2f
            directionalLight = DirectionalLight(2.0, 2.0, 2.0) // set the direction
            directionalLight!!.power = 1f
            mPicker = ObjectColorPicker(this)
            mPicker?.setOnObjectPickedListener(this)
            loaderOBJ = LoaderOBJ(this, currentObject3D.get().hashCode())
            restoreStatus.postValue(true)
            val cx: Int = viewportHeight
            val cy: Int = viewportWidth
            val parsedObj = loaderOBJ?.parse(true)
            object3D = parsedObj?.parsedObject
            val finalRadius = sqrt((cx * cx + cy * cy).toDouble()).toFloat()
            val xPos = finalRadius.times(0.000010f).toDouble()
            val yPos = finalRadius.times(-0.00028f).toDouble()
            object3D?.setPosition(xPos, yPos, 0.0)
            val realScale = object3D?.scaleX?.times(0.010f)
            object3D?.setScale(realScale!!)
            listChildren(object3D!!)
            currentCamera.farPlane = 10.0
            currentCamera?.lookAt = Vector3(0.0, 0.0, 0.0)

            newScene.addLight(directionalLight)
            newScene.addLight(mLight)
            newScene.addChild(object3D)
            newScene.backgroundColor = context.getColor(android.R.color.holo_blue_bright)
            addAndSwitchScene(newScene)

        } catch (e: ParsingException) {
            Log.e("TAG", e.toString())
            e.printStackTrace()
        }
    }
```
  binding?.start?.setOnClickListener {
            renderer?.currentObject3D?.set(R.raw.obj_skeleton_selectable)
            renderer?.initialize(newScene = Scene(renderer))
        }
```

**Comment by contriteobserver** [comment]
It is not obvious, but we have two threads implicit in this case, a UI thread and a Render thread.

Initializing a scene in the UI thread is problematic, a more straightforward approach would be to initialize both scenes in `initScene`, and then call `Renderer.switchScene()` in the UI thread. FYI `swichScene` is deferred, see `SwitchSceneDirect`, until the next available `onRenderFrame`.

**Comment by jonathanchh1** [comment]
Hey, thanks for getting back to me. Hmm, could you share a code snippet as to how you would go about achieving this, and everything one wants to switch to previous scenes, how can that be achieved? I suspect scene caching is also needed to derive at that?

**Comment by contriteobserver** [comment]
Here is a `1.2.1970` example:
```
public class MainActivity extends AppCompatActivity {
    SurfaceView view;
    SceneSwitchingRenderer renderer;
    SharedPreferences mPrefs;
    int mActiveScene;
    static final int SCENE_INDEX = 1;
    static final String ACTIVE_SCENE = "active_scene";

    @SuppressLint("ClickableViewAccessibility")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mPrefs = getSharedPreferences("unique_preference_key", Context.MODE_PRIVATE);
        mActiveScene = mPrefs.getInt(ACTIVE_SCENE, SCENE_INDEX);

        renderer = new SceneSwitchingRenderer(getApplicationContext());
        view = findViewById(R.id.glsl_content);
        view.setSurfaceRenderer(renderer);
        view.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View view, MotionEvent motionEvent) {
                renderer.onTouchEvent(motionEvent);
                return false;
            }
        });
    }
    
    @Override
    protected void onPause() {
        super.onPause();
        renderer.onPause();

        SharedPreferences.Editor ed = mPrefs.edit();
        ed.putInt(ACTIVE_SCENE, mActiveScene);
        ed.commit();
    }

    @Override
    protected void onResume() {
        super.onResume();
        renderer.onResume();
    }

    class SceneSwitchingRenderer extends Renderer {
        Scene altScene;

        public SceneSwitchingRenderer(Context context) {
            super(context);
        }

        @Override
        protected void initScene() {
            try {
                getCurrentScene().setBackgroundColor(Color.CYAN & Color.DKGRAY);

                ALight key = new DirectionalLight(1,-1,-1);
                getCurrentScene().addLight(key);

                Material material = new Material();
                material.setColor(Color.YELLOW & Color.GRAY);
                material.setDiffuseMethod(new DiffuseMethod.Lambert());
                material.setSpecularMethod(new SpecularMethod.Phong());
                material.setAmbientColor(Color.CYAN | Color.GRAY);
                material.setAmbientIntensity(1/8d,1/8d, 1/8d);
                material.enableLighting(true);

                Object3D obj = new NPrism(32,0, 1,2);
                obj.setMaterial(material);
                getCurrentScene().addChild(obj);

                getCurrentCamera().setPosition(5,4,3);
                getCurrentCamera().setLookAt(obj.getPosition());
            } catch (Exception e) {
                Log.e(getLocalClassName() + ".initDefaultScene", e.getMessage());
            }

            try {
                altScene = new Scene(this);
                altScene.setBackgroundColor(Color.MAGENTA & Color.DKGRAY);

                ALight key = new DirectionalLight(-1,-1,-1);
                altScene.addLight(key);

                Material material = new Material();
                material.setColor(Color.RED & Color.GRAY);
                material.setDiffuseMethod(new DiffuseMethod.Lambert());
                material.setSpecularMethod(new SpecularMethod.Phong());
                material.setAmbientColor(Color.MAGENTA | Color.GRAY);
                material.setAmbientIntensity(1/8d,1/8d, 1/8d);
                material.enableLighting(true);

                Object3D obj = new Torus(2,1,32,12);
                obj.setMaterial(material);
                altScene.addChild(obj);

                altScene.getCamera().setPosition(8,7,6);
                altScene.getCamera().setLookAt(obj.getPosition());

                if(mActiveScene ==0) {
                    addScene(altScene);
                } else {
                    addAndSwitchScene(altScene);
                }
            } catch (Exception e) {
                Log.e(getLocalClassName() + ".initAltScene", e.getMessage());
            }
        }

        @Override
        public void onOffsetsChanged(float xOffset, float yOffset, float xOffsetStep, float yOffsetStep, int xPixelOffset, int yPixelOffset) {

        }

        @Override
        public void onTouchEvent(MotionEvent event) {
            switch(event.getAction()) {
                case MotionEvent.ACTION_DOWN:
                    if (renderer.getCurrentScene() == altScene) {
                        renderer.switchScene(0);
                        mActiveScene = 0;
                    } else {
                        renderer.switchScene(altScene);
                        mActiveScene = 1;
                    }
                    break;
            }
        }
    }
}
```
![twoscene-2021-02-25](https://user-images.githubusercontent.com/17471201/109190844-632b4f00-774a-11eb-85c7-cf96c66c17ef.gif)

**Comment by jonathanchh1** [comment]
Thanks for the snippet, i tried it and it didn't really work. still crashes after loading the second model. i see in your example you are not using the LoaderOBJ but just creating an object. how about loading two or three humanoid models using the loader?

**Comment by contriteobserver** [comment]
That is interesting, I confirmed expected behavior of this snippet on a number of devices before posting, please confirm which specific devices present this unexpected behavior. Please also confirm the version exhibiting this behavior.

"After loading the second model" is a bit vague, are we seeing an exception in `initScene` or is the crash somewhere else? If in a try block, changing the exception to `Exception.printStackTrace()` might be revealing.

Unfortunately, I do not have access to two or three hundred different humanoid models, so I  cannot be of help there.

**Comment by jonathanchh1** [comment]
SeeSo, there's no exception being returned, it literally stops after returning garbage collection message like this:

```
Parsing group: Muscle_Neck
parsing object: Object_Muscle_Neck1060694241
Background concurrent copying GC freed 240839(10034KB) AllocSpace objects, 2(92KB) LOS objects, 63% free, 28MB/76MB, paused 79us total 

- [ ] 104.245ms

System: A resource failed to call release. 
```
Thats what it shows me and nothing else then it crashes. i think it might just have to do with the complexity of the object being loaded..meaning texture or other things.. not completely sure, and i am talking about loading one to three objects maybe four of file size 4 to 10 mbs

**Comment by contriteobserver** [comment]
Resource exhaustion is a possibility considering the file sizes, the management of resources is what makes mobile development so interesting. Is the application able to run with simpler models?

Wavefront OBJ file formats permits polygons of any number of sides, but Rajawali can only process triangles and quads correctly. Do any of the anatomy models contain polygons of more than 4 sides?

**Comment by jonathanchh1** [comment]
yes, the application is able to run with simpler models. someone else produces the files for me therefore, i had no control over the size of the file or the initial scale of the model and i believe it was life-size previously but once, i received a new scaled down version at 15% scaled. I was able to load multiple models but i figured out a solution, i am now caching the models after first initial loaderObj call, so, continous calls aren't necessary and enabled scene-caching to allow for easier switch. 

Okay, this isn't an issue but quick question, is there away to associate an image programmatically with the mtl file without adding it on the drawable folder? and can you pass me a sample code if you don't mind. thank you again.

**Comment by contriteobserver** [comment]
First off, there is a 3D graphics technique of [Baking Normal Maps](https://en.wikipedia.org/wiki/Normal_mapping) which allows one to maintain detail while simplifying mesh geometry. Asset Optimization is all part of the workflow of making a production app.

Second, avoid putting textures in the `drawable` folder, instead use the `drawable-nodpi` directory. Android instantiates multiple scaled versions of images in the `drawable` folder, but does not scale images in `drawable-nodpi` a real memory saver.

I never put OBJ or MTL files in the `drawable` folder, instead I use the `raw` folder, that works by default.

**Comment by jonathanchh1** [comment]
So, i am using the raw folder to hold both mtl and raw files however, i want to add the mtl file programmatically and the same for the drawable images.

**Comment by contriteobserver** [comment]
Not sure I understand correctly, `LoaderOBJ.parse()` currently extracts only a limited amount color and texture information from MTL files, it is however possible to programmatically:

- load a image into a texture using `new Texture("texture name", R.drawable.image)`
- add a texture to a material with `Material.addTexture(texture)`
- add a material to an object with `Object3D.setMaterial(material)`

Typically one just calls `Object3D.clone()` on the parsed object, programmatically updates the materials as needed, and adds the new object into the scene. This approach also provides access to dynamic materials and animated textures, which lie outside the scope of MTL constructs.

**Comment by jonathanchh1** [comment]
oh okay, what i am trying to say is, is there away to load an obj file with its mtl file from a server dynamically using the library?

**Comment by contriteobserver** [comment]
Unfortunately, not yet. The issue is larger than OBJ and MTL loading, there are also texture files. Trying to get a decent integration with [okhttp](https://square.github.io/okhttp/) implemented, but we are not there yet.

Best workaround we have right now is to download the files from a server to the android filesystem, and then load them into Rajawali from there.

Will see what can be done, addressed a small part of the issue with  #2313

**Comment by jonathanchh1** [comment]
oh okay. thank you so much! i will go with that solution and i'll close out this issue.
